consensus/service/src/byzantine_ledger/worker.rs

```diff
@@ -483,120 +483,111 @@ impl<
+            .expect("failed appending block");
 
-                block.id
-            };
-
-            // Sanity check + update current slot.
-            let current_slot_index =
-                self.ledger.num_blocks().expect("num blocks failed") as SlotIndex;
+        counters::TX_EXTERNALIZED_COUNT.inc_by(externalized.len() as i64);
 
+        // Update current slot index.
+        self.current_slot_index = {
+            let current_slot_index: SlotIndex = self.ledger.num_blocks().unwrap();
             assert_eq!(current_slot_index, self.current_slot_index + 1);
+            current_slot_index
+        };
 
-            self.current_slot_index = current_slot_index;
-            self.prev_block_id = block_id;
-
-            // Purge transactions that can no longer be processed based on their tombstone block.
-            let max_externalized_slots = self.scp.max_externalized_slots() as u64;
-            let purged_hashes = {
-                let index = current_slot_index.saturating_sub(max_externalized_slots);
-                self.tx_manager.remove_expired(index)
-            };
-
-            // Drop pending values that are no longer considered valid.
-            let tx_manager = self.tx_manager.clone();
-            self.pending_values_map.retain(|tx_hash, _| {
-                !purged_hashes.contains(tx_hash) && tx_manager.validate(tx_hash).is_ok()
-            });
-            // help the borrow checker
-            let self_pending_values_map = &self.pending_values_map;
-            self.pending_values
-                .retain(|tx_hash| self_pending_values_map.contains_key(tx_hash));
-
-            debug_assert!(self.pending_values_map.len() == self.pending_values.len());
-
-            log::info!(
-                self.logger,
-                "number of pending values post cleanup: {} ({} expired)",
-                self.pending_values.len(),
-                purged_hashes.len(),
-            );
+        // Purge transactions that can no longer be processed based on their tombstone block.
+        let max_externalized_slots = self.scp_node.max_externalized_slots() as u64;
+        let purged_hashes = {
+            let index = self
+                .current_slot_index
+                .saturating_sub(max_externalized_slots);
+            self.tx_manager.remove_expired(index)
+        };
 
-            // Previous slot metrics.
-            counters::PREV_SLOT_NUMBER.set((current_slot_index - 1) as i64);
-            counters::PREV_SLOT_ENDED_AT.set(chrono::Utc::now().timestamp_millis());
-            counters::PREV_SLOT_NUM_EXT_VALS.set(ext_vals.len() as i64);
-            counters::PREV_NUM_PENDING_VALUES.set(self.pending_values.len() as i64);
+        // Drop pending values that are no longer considered valid.
+        let tx_manager = self.tx_manager.clone();
+        self.pending_values_map.retain(|tx_hash, _| {
+            !purged_hashes.contains(tx_hash) && tx_manager.validate(tx_hash).is_ok()
```

Comment:
it seems odd to check both conditions... what does `self.tx_manager.remove_expired(index)` do above?

Reply:
`self.tx_manager.remove_expired(index)` removes transactions from the cache once their tombstone block has been exceeded. The code here removes pending transactions that are no longer viable, so that they won't get into SCP.